Background Jobs Work

ikreymer opened the pull request:
Fixes #1252 

To be merged after #1320, builds on top of that, merging initial background jobs work

This now supports a CreateReplicaJob and DeleteReplicaJob. CreateReplicaJob runs on new crawls, uploads, profiles and updates the `replicas` array with the info about the replica after the job succeeds.
DeleteReplicaJob deletes the replica.
Both jobs are created from the new `replica_job.yaml` template. The CreateReplicaJob sets secrets for primary storage + replica storage, while DeleteReplicaJob only needs the replica storage.

The job is processed in the operator when the job is finalized (deleted), which should happen immediately when the job is done, either because it succeeds or because the backoffLimit is reached (currently set to 3).

Some potentially questions:
- Should there be a limit to number of retries, or should try to retry indefinitely if replication or deletion of replication fails? The failure is logged in the db after the job is removed.
- Should the deletion be paused if deleting a replica fails? It seems like it should not, though the object is removed even if deleting the replica fails..

tw4l committed:
WIP: First pass at background jobs to handle replicas

Not yet integrated with work to support multiple storages

ikreymer committed:
fixes to get replicationjobs running!
- fix various paths in replication job / consistent naming
- add endpoint urls passed in to rclone config
- currently need to strip off bucket from endpoint_url and add to file path
- supports creating multiple replicas, as many as configured for org / default storages!
- enable minio port for debugging

ikreymer committed:
more typing fixes:
- move crawl states to models
- import typed ops to in ops classes, whereever possible while avoiding circular imports
- fix various typing issues
- removing optional from models where unnecessary

ikreymer committed:
uuid consistency pass:
use 'UUID' instead of UUID4 or uuid.UUID for UUID types

ikreymer committed:
Merge branch 'uuid-consistency' into more-typing-fixes

ikreymer committed:
fix merge

ikreymer committed:
operator type fixes

ikreymer committed:
more type fixes from check-untyped-defs

ikreymer committed:
more stringent type checking, mypy --check-untyped-defs should now pass

ikreymer committed:
cleanup base crawls / crawls / uploads to use consistent init arg org order

ikreymer committed:
add more type hints wrapped in TYPE_CHECKING conditional to avoid circular import

ikreymer committed:
fix type check conditional

ikreymer committed:
states: tuple -> list

ikreymer committed:
readd missing api

ikreymer committed:
mark pod resource values as non-optional

ikreymer committed:
Merge branch 'more-typing-fixes' into bg-jobs-2

ikreymer committed:
additional changes for bg-jobs

ikreymer committed:
fix

ikreymer committed:
refactor

ikreymer committed:
fix arg typo

ikreymer committed:
replication job handling:
- keep track of file that replicatejob references
- update crawlfile and profilefile replicas after job succeeds
- handle replicatejob failures
- move job handling to finalize method
- remove generic update job for now

ikreymer committed:
add priority for background job
always finalize, unless db update exception occurred

ikreymer committed:
always set 'finished' time

ikreymer committed:
support deletion of replicas:
- use BgJobType Enum, 'create-replica' and 'delete-replica' to refer to different replica jobs
- unify under single job template in replica_job.yaml
- support special handling (updating files) after create-replica job, none for delete-replica

ikreymer committed:
minio: add 'minio_local_console_port' option to enable mapping minio console 9001 to a local port for debugging

tw4l committed:
Merge branch 'main' into issue-1252-background-jobs-latest

tw4l committed:
Set default replica location and test it

tw4l committed:
Fix resources reference for upload test

tw4l committed:
Check numReplicas, not replicas (not in CrawlFileOut)

tw4l committed:
Flesh out replication tests

tw4l committed:
Update test.yaml storages

tw4l committed:
Wait for background job to finish

tw4l committed:
Return BackgroundJob from GET endpoint

tw4l committed:
Filter jobs by type

tw4l committed:
Add missing return

ikreymer committed:
tests: create 'replica-0' bucket for testing, fix test storage path

tw4l committed:
Fixups

tw4l committed:
Add additional background_jobs tests

tw4l committed:
Return specific background job model in GET endpoint

tw4l committed:
Verify replicated files directly from minio and fix tests

ikreymer commented on backend/btrixcloud/templates/redis.yaml:
```diff
@@ -110,7 +110,7 @@ spec:
 
       readinessProbe:
-        initialDelaySeconds: 10
+        initialDelaySeconds: 0
```
```suggestion
        initialDelaySeconds: 10
```

ikreymer commented on backend/btrixcloud/templates/redis.yaml:
```diff
@@ -119,7 +119,7 @@ spec:
 
       livenessProbe:
-        initialDelaySeconds: 10
+        initialDelaySeconds: 0
```
```suggestion
        initialDelaySeconds: 10
```

ikreymer committed:
undo unrelated changes in redis.yaml

tw4l committed:
One more go

tw4l committed:
Fix test for GET /jobs/{job_id}

tw4l committed:
Fix typo

tw4l committed:
Use strs for type comparison

ikreymer committed:
add AnyJob Union type for response model, otherwise only base type is matched
get_background_job() returns specific job type
reformat with black

tw4l committed:
Fix typo

tw4l committed:
Wait 10 seconds before checking mongo

tw4l committed:
Expect original file and replica to be deleted

tw4l committed:
Modify return types

tw4l committed:
Just print log line if file deleted before replication

tw4l committed:
Not an fstring

tw4l committed:
Be extra careful

tw4l committed:
Don't expect failed replication jobs

tw4l committed:
Print log hashes

tw4l committed:
Use temp.name for filepath to temp file

tw4l committed:
Tweak tempfile and add print logging

tw4l committed:
Temporarily remove exception handling to see what issue is

tw4l committed:
One step closer

tw4l committed:
Get rid of option from Python 3.12

tw4l committed:
Remove leftover colon

tw4l committed:
Add back try/except

tw4l committed:
One more go with print logging

tw4l committed:
Import os

tw4l committed:
Print log file path

tw4l committed:
Remove os import

tw4l committed:
Fix path to files to verify replication

tw4l committed:
Format with black

tw4l committed:
Get endpoint url for minio:30090 from env var

tw4l committed:
Remove port from microk8s HOST_URL_FROM_K8S

tw4l committed:
Remove try/except again

tw4l committed:
Add os import

tw4l committed:
Add back try/except

tw4l committed:
Strip primary bucket from replica path

ikreymer committed:
use 127.0.0.1:30090 for minio access

tw4l committed:
Remove try/except again

ikreymer committed:
keep one service for minio access

ikreymer committed:
export port range

ikreymer committed:
try individual ports

ikreymer committed:
fix password, simplify downloading via get_bucket

tw4l committed:
Test for NoSuchKey exception

tw4l committed:
Import exceptions from botocore differently

tw4l committed:
One more go

tw4l committed:
Add operator print logging for CrawlFile to compare path

tw4l committed:
One more go at print logging paths

tw4l committed:
Print log path written to CreateReplicaJob

tw4l committed:
Strip btrix-test-data from job filepath for both primary and replica if present

tw4l committed:
Format

tw4l committed:
Add another time.sleep and more print logging

tw4l committed:
Add test to ensure there are no failures

tw4l committed:
Wait full minute

tw4l committed:
Print log filename from CrawlCompleteIn

tw4l committed:
Hail mary

tw4l committed:
Fix crawler_image typo in test.yaml

ikreymer committed:
various fixes:
- ensure DeleteReplicaJob only stores relative path, not bucket, to be consistent
- jobs api: return specific job type object in list
- tests: ensure buckets are not part of file_path
- tests: fix replica test by changing sort order, don't test replica of file that was deleted!

ikreymer committed:
switch assert to print failed jobs

ikreymer committed:
bump resources for replica jobs

ikreymer committed:
:x

tw4l committed:
Remove print logging

tw4l committed:
Typo

tw4l commented:
> Some potentially questions:
> 
>     * Should there be a limit to number of retries, or should try to retry indefinitely if replication or deletion of replication fails? The failure is logged in the db after the job is removed.

The current limit of 3 tries with increasing backoff seems reasonable to me.

>     * Should the deletion be paused if deleting a replica fails? It seems like it should not, though the object is removed even if deleting the replica fails..

Hm, I'm also tempted to say it should not. But maybe we should have a way of flagging when background jobs fail in the UI. As a follow-up, could add the list of background jobs to the Org Settings tab for org owners, easily filterable on failures?

Another follow-up might be an easy way to retry a failed job, similar to what we have for event webhook notifications.

ikreymer committed:
jobs: derived background job names from object-ids for clarity
reduce print() statements
tests: undo debug test changes, update image

ikreymer committed:
fix bucket create call, also test on microk8s

ikreymer committed:
also check nightly

ikreymer committed:
force test

ikreymer committed:
derive unique job_id from job_id_prefix

ikreymer committed:
oops, don't disable port!

ikreymer committed:
tests: move replica tests to nightly tests only
- revert test changes for k3d / microk8s
- add test_upload_replicas to nightly to test replicas after upload
- add create replicas test to test_crawl_timeout
- add delete replicas test to test_delete_crawls
- add test_z_background_jobs to nightly
- increase page limit, lower crawl stop timeout to avoid crawls finished too quickly!
- add test-nightly-addons.yaml for configuring replicas

ikreymer committed:
undo changes for testing, revert running nightly on push

ikreymer committed:
bump test to crawler 0.12.0
